hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+
+- Level 1 (leaves): The existing chunk level checksums (written by the client and verified by the existing datanode container scanner).
+- Level 2: A block level checksum created by hashing all the chunk checksums within the block.
+- Level 3 (root): A container level checksum created by hashing all the block checksums within the container.
+    - This top level container hash is what is reported to SCM to detect diverged replicas. SCM does not need block or chunk level hashes.
+
+When SCM sees that replicas of a non-open container have diverged container checksums, it can trigger a reconciliation process on all datanode replicas. SCM does not need to know which container hash is correct (if any of them are correct), only that all containers match. Datanodes will use their merkle tree and those of the other replicas to identify issues with their container. Next, datanodes can read the missing data from existing replicas and use it to repair their container replica.
+
+### Phase I (outlined in this document)
+
+- Add container level checksums that datanodes can compute and store.
+
+- Add a mechanism for datanodes to reconcile their replica of a container with another datanode's replica so that both replicas can be verified to be equal at the end of the process.
+
+- Add a mechanism for SCM to trigger this reconciliation as part of the existing heartbeat command protocol SCM uses to communicate with datanodes.
+
+- Add an `ozone admin container reconcile <container-id>` CLI that can be used to manually resolve diverged container states among non-open container replicas.
+
+### Phase II (out of scope for this document)
+
+- Automate container reconciliation requests as part of SCM's replication manager.
+
+- Simplify SCM replication manager decommission and recovery logic based on mismatch of container checksums, instead of the combination of all possible container states.
+
+### Phase III (out of scope for this document)
+
+- Extend container level checksum verification to erasure coded containers.
+    - EC container replicas do not have the same data and are not expected to have matching container checksums.
+    - EC containers already use offline recovery as a reconciliation mechanism.
+
+## Solution Implementation
+
+### Storage
+
+The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):
+
+```
+ContainerChecksum: {
+  algorithm: CRC32
+  checksum: 12345
+  repeated BlockChecksum
+}
+BlockChecksum: {
+  algorithm: CRC32
+  checksum: 12345
+  length: 5
+  deleted: false
```

Comment:
We currently assume that there is no block deletion by mistake as the path for deletion of block comes from OM receiving block deletion. Thus, a block is marked as deleted only when OM has deleted the object and the blocks referenced. Protection against accidental deletes should be done via snapshots.

Reply:
The might not be a valid assumption since, when there is a corruption, it is invalid to assume "something will never be corrupted".  A possible cause of block deletion by mistake is a (future) software bug.

We could have another checksum for the deleted blocks.  Below is a quick idea:
1. the deleted-block-checksum `sum` initially is 0;
2. When a block `b` is deleted, update `sum` with xor, i.e. `sum ^= hash(b)`.

SCM could store a deleted-block-checksum for each container.  Then, it is easy to tell if the deleted blocks are correct in a particular container.

Reply:
A couple of issues
1. The flag `deleted` in code should be true only if SCM issued the deletion.
2. SCM tracking a lineage of hash changes has other memory bloat challenges for SCM which will provide fiction for scaling.
3. If SCM has a software bug that erroneously sends deletes or Datanode erroneously marks blocks as deleted, protection via MT might not be enough.
4. The only way to make sure that during reconciliation we know for sure that a block is deleted at the SCM level (not just peer) would introduce a lot of bloat to SCM.

There are invariants in this design that depend on correctness of implementation
1. Two chunks written by Ratis to two Datanodes cannot do both
    1. Agree with the checksum provided by the client
    2. Disagree with the peer Datanode on the checksum
2.  SCM decides when a block is deleted and it informs the Datanodes (followers) on the decision to delete a block. This is the only path to delete a block in a Datanode and for it to be marked as deleted on the Datanode. 
    1. If SCM sends a delete command to one Datanode (follower) it is ok for the other followers to learn about it via peer to peer communication.

Reply:
> 2. SCM tracking a lineage of hash changes has other memory bloat challenges for SCM which will provide fiction for scaling.

This is just xor checkout sum over block id hash (not block data hash) for deleted blocks.  The overhead should be similar to an integer deleted block count.

> 3. ...  SCM has a software bug ...

A software bug could be in datatnodes: one datanode has mistakenly marked a block as deleted.  SCM reconciles the container with other datanodes.  During reconciliation, the other datanodes will also make that block as deleted.

Reply:
Not sure I follow the recommendation for how to track deletions beyond adding a flag to indicate it is deleted or how the guarantees would be more robust.

Reply:
We have two checksums `mtChecksum` and `deletedBlockChecksum`.  When SCM issue a reconciliation, it sends both the checksums to the datanodes:
- When the `mtChecksum` is matched, it means all the block data are correct.
  - `mtChecksum` mismatched means that some data is corrupted.
- When the `deletedBlockChecksum` is matched, it means the deleted block list is good.
   - `deletedBlockChecksum ` mismatched means that the deleted block list is incorrect.

Without `deletedBlockChecksum`, `mtChecksum` matched means both all the block data correct and deleted block list correct since it won't verify the block list.

Reply:
Why split the logic? If a Container checksum does not match
1. Check if the container have the same blocks
    1. If blocks differ fix blocks
2. Check if any blocks have been deleted on the peer.

The additional checksum to track the sum of deleted blocks does not lead to any simplification of the runtime processing and adds a fork in the logic at the root vs. down the tree on a per block basis.

Reply:
> Why split the logic? If a Container checksum does not match

It is for the other case.  If we don't split it, the deleted list could be different when container checksums are matched.

Reply:
It could hit the problem described in https://github.com/apache/ozone/pull/6121#discussion_r1488638589

> A software bug could be in datatnodes: one datanode has mistakenly marked a block as deleted. SCM reconciles the container with other datanodes. During reconciliation, the other datanodes will also make that block as deleted.